### ISSUE TITLE
Bug with computeInputScript and taproot outputs

### ISSUE THREAD

**Issue by benthecarman** [lightningnetwork/lnd#6678]
### Background

Signing a taproot input with `computeInputScript`

### Your environment

* version of `lnd` 

0.15.0

### Steps to reproduce

When calling `computeInputScript` if you set the `output` in the `SignDescriptor` it must be excluded from the `prevOutputs` otherwise it will create an invalid schnorr signature. This makes signing multiple inputs with a single call to `computeInputScript` unworkable.

### Expected behaviour

It should filter out the `output` from `prevOutputs` when signing that input

### Actual behaviour

Creates an invalid schnorr signature

**Comment by benthecarman** [comment]
Looking into this more, seems like it actually still produces an invalid schnorr signature. Not sure what I am doing wrong

**Comment by alexbosworth** [comment]
Did you try using `SignOutputRaw`

**Comment by guggero** [comment]
What's your exact use case? Are you trying to spend an output that belongs to the wallet? The `prevOutputs` needs to contain _all_ previous outputs, otherwise things won't work.

**Comment by guggero** [comment]
Actually, `ComputeInputScript` ignores the `prevOutputs` completely and cannot be used to sign for p2tr inputs (at least not over the RPC). The correct method to use is `SignOutputRaw` or the `FundPSBT`/`FinalizePSBT` methods.

**Comment by benthecarman** [comment]
> Did you try using SignOutputRaw

Yes I get the same issue. I've tried every different variation of hash type and sign method and always result in an invalid signature

> Actually, ComputeInputScript ignores the prevOutputs completely and cannot be used to sign for p2tr inputs (at least not over the RPC).

That would be nice to document or fix because the current documentation suggests it can

>  The correct method to use is SignOutputRaw or the FundPSBT/FinalizePSBT methods.

Trying with `SignOutputRaw` but still get same errors. FundPSBT/FinalizePSBT are usable for me here because the transaction is being constructed by a third party

**Comment by alexbosworth** [comment]
I can confirm that SignOutputRaw works but invalid signature can definitely be a tough one to debug from and there are various types of taproot signature types which can make it tricky too

**Comment by guggero** [comment]
> That would be nice to document or fix because the current documentation suggests it can

What documentation do you mean? https://api.lightning.community/#computeinputscript only mentions p2wkh and nested p2wkh.

What parameters are you passing to `SignOutputRaw`? Are you trying to do BIP86 key spend? Does the wallet know the BIP86 address? Or only the raw private key?

**Comment by alexbosworth** [comment]
Sorry I'm actually not sure if it works for internal key space keys so I don't have an example if that's what you are referring to

However I don't know why you couldn't use PSBT in that case?

**Comment by benthecarman** [comment]
> What parameters are you passing to SignOutputRaw? Are you trying to do BIP86 key spend? Does the wallet know the BIP86 address? Or only the raw private key?

Yes I am receiving to the LND wallet in a taproot address then trying to spend from it. I am using `SIGN_METHOD_TAPROOT_KEY_SPEND_BIP0086`, giving the prevOuts in order, and using the `output` to sign it (instead of key locator)

> However I don't know why you couldn't use PSBT in that case?

This makes it super ugly where everytime I want to sign I need to create a dummy psbt, call fund psbt, and copy the input records to the real psbt so I can call `signpsbt` properly

**Comment by alexbosworth** [comment]
Ah yes it isn't the most beautiful method but that is the only way to get the derivation paths and pass them as far as I know

**Comment by benthecarman** [comment]
Here is what I am doing if you are curious: https://github.com/bitcoin-s/bitcoin-s/pull/4428/files

**Comment by alexbosworth** [comment]
Would it work for your use case if you receive to a P2TR address that isn't in the main wallet keys?

**Comment by benthecarman** [comment]
> Would it work for your use case if you receive to a P2TR address that isn't in the main wallet keys?

Not really 

> `HashType.sigHashAll,` definitely needs to be `SigHashDefault`

Does not fix :(

**Comment by guggero** [comment]
You're not passing in any key descriptor, so it's going to use family 0 and index 0 which isn't the key you want to use. Can you control what address the external wallet sends the funds to? Because then I'd first derive a key (so you know the family and index), create an address for that to receive and then later specify that key family and index when signing.

**Comment by guggero** [comment]
BTW, this is probably the best documentation of the API right now: https://github.com/lightningnetwork/lnd/blob/master/lntest/itest/lnd_taproot_test.go#L275

**Comment by benthecarman** [comment]
I see, so is there no way to do it for outputs internal to the wallet at the moment?

**Comment by alexbosworth** [comment]
I would use FundPsbt/SignPsbt that should work for P2TR

**Comment by guggero** [comment]
> I see, so is there no way to do it for outputs internal to the wallet at the moment?

No, not without using `FundPSBT`.

**Comment by guggero** [comment]
`SignOutputRaw` isn't built to deal with wallet outputs, so the title should be changed to `ComputeInputScript` only.

But then you'd call `ComputeInputScript` with the full TX that has all inputs and outputs present?

**Comment by benthecarman** [comment]
Yeah that is how I was doing it before

https://github.com/ln-vortex/ln-vortex/blob/master/lnd/src/main/scala/com/lnvortex/lnd/LndVortexWallet.scala#L107

**Comment by guggero** [comment]
Can you please check if you get things working with #6680?
